handlers/macrokiosk/macrokiosk.go

```diff
@@ -1,5 +1,227 @@
-*/
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/buger/jsonparser"
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/gsm7"
+	"github.com/nyaruka/courier/handlers"
+	"github.com/nyaruka/courier/utils"
+	"github.com/nyaruka/gocommon/urns"
+)
+
+var (
+	sendURL                   = "https://www.etracker.cc/bulksms/send"
+	maxMsgLength              = 1600
+	configMacrokioskSenderID  = "macrokiosk_sender_id"
+	configMacrokioskServiceID = "macrokiosk_service_id"
+)
+
+func init() {
+	courier.RegisterHandler(newHandler())
+}
+
+type handler struct {
+	handlers.BaseHandler
+}
+
+func newHandler() courier.ChannelHandler {
+	return &handler{handlers.NewBaseHandler(courier.ChannelType("MK"), "Macrokiosk")}
+}
+
+// Initialize is called by the engine once everything is loaded
+func (h *handler) Initialize(s courier.Server) error {
+	h.SetServer(s)
+	err := s.AddHandlerRoute(h, http.MethodPost, "status", h.receiveStatus)
+	if err != nil {
+		return err
+	}
+	err = s.AddHandlerRoute(h, http.MethodGet, "status", h.receiveStatus)
+	if err != nil {
+		return err
+	}
+	err = s.AddHandlerRoute(h, http.MethodGet, "receive", h.receiveMessage)
+	if err != nil {
+		return err
+	}
+	return s.AddHandlerRoute(h, http.MethodPost, "receive", h.receiveMessage)
+}
+
+type statusForm struct {
+	MsgID  string `name:"msgid" validate:"required"`
+	Status string `name:"status" validate:"required"`
+}
+
+var statusMapping = map[string]courier.MsgStatusValue{
+	"ACCEPTED":    courier.MsgSent,
+	"DELIVERED":   courier.MsgDelivered,
+	"UNDELIVERED": courier.MsgFailed,
+	"PROCESSING":  courier.MsgWired,
+}
+
+// receiveStatus is our HTTP handler function for status updates
+func (h *handler) receiveStatus(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.Event, error) {
+	form := &statusForm{}
+	err := handlers.DecodeAndValidateForm(form, r)
+	if err != nil {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, err)
+	}
+
+	msgStatus, found := statusMapping[form.Status]
+	if !found {
+		return nil, courier.WriteAndLogRequestIgnored(ctx, w, r, channel, fmt.Sprintf("ignoring unknown status '%s'", form.Status))
+	}
+	// write our status
+	status := h.Backend().NewMsgStatusForExternalID(channel, form.MsgID, msgStatus)
+	err = h.Backend().WriteMsgStatus(ctx, status)
+	if err == courier.ErrMsgNotFound {
+		return nil, courier.WriteAndLogStatusMsgNotFound(ctx, w, r, channel)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return []courier.Event{status}, courier.WriteStatusSuccess(ctx, w, r, []courier.MsgStatus{status})
+
+}
+
+type moForm struct {
+	Longcode  string `name:"longcode"`
+	Shortcode string `name:"shortcode"`
+	MSISDN    string `name:"msisdn"`
+	From      string `name:"from"`
+	Text      string `name:"text"`
+	MsgID     string `name:"msgId"`
+	Time      string `name:"time"`
```

Comment:
Fields I see in our logs:

`GET /handlers/macrokiosk/receive/1f5c4697-16a0-4b5e-b184-07b40a76e9fb?from=60179085717&text=Pada+pendapat+saya%2c+kita+sebagai+masyarakat+mestilah+bekerjasama+dengan+pihak+sekolah+untuk+mengurangkan+kes+buli+yang+berlaku+di+sekolah.&time=2018-02-2316%3a05%3a00&shortcode=69898&telcoid=1&msgid=180223160500493421`

So only need from, text, time and msgid